Source/Renderer/Context.js

```diff
@@ -1028,23 +1061,28 @@ define([
     Context.prototype.readPixels = function(readState) {
         var gl = this._gl;
 
-        readState = readState || {};
-        var x = Math.max(readState.x || 0, 0);
-        var y = Math.max(readState.y || 0, 0);
-        var width = readState.width || gl.drawingBufferWidth;
-        var height = readState.height || gl.drawingBufferHeight;
+        readState = defaultValue(readState, defaultValue.EMPTY_OBJECT);
+        var x = Math.max(defaultValue(readState.x, 0), 0);
+        var y = Math.max(defaultValue(readState.y, 0), 0);
+        var width = defaultValue(readState.width, gl.drawingBufferWidth);
+        var height = defaultValue(readState.height, gl.drawingBufferHeight);
         var framebuffer = readState.framebuffer;
 
         //>>includeStart('debug', pragmas.debug);
         Check.typeOf.number.greaterThan('readState.width', width, 0);
         Check.typeOf.number.greaterThan('readState.height', height, 0);
         //>>includeEnd('debug');
 
-        var pixels = new Uint8Array(4 * width * height);
+        var pixelDatatype = PixelDatatype.UNSIGNED_BYTE;
+        if (defined(framebuffer) && framebuffer.numberOfColorAttachments > 0) {
+            pixelDatatype = framebuffer.getColorTexture(0).pixelDatatype;
+        }
+
+        var pixels = PixelFormat.createTypedArray(PixelFormat.RGBA, pixelDatatype, width, height);
```

Comment:
`PixelFormat.createTypedArray` - this function may need to handle half float.

Reply:
It does. The array buffer for a half float is a `Uint16Array`. It's awkward, but there is no `Float16Array`.

Reply:
Ah, so `readPixels` expects a `Uint16Array` instead of a `Float32Array` for half float?

Reply:
Yes, same with `texImage2D` and `texSubImage2D`. You can't have a different `format` and `internnalFormat` until WebGL 2.